src/auto_renewal.cairo

```diff
@@ -0,0 +1,242 @@
+trait IAutoRenewal<TContractState> {
+    fn is_renewing(
+        self: @TContractState,
+        domain: felt252,
+        renewer: starknet::ContractAddress,
+        limit_price: u256
+    ) -> felt252;
+
+    fn get_contracts(
+        self: @TContractState
+    ) -> (starknet::ContractAddress, starknet::ContractAddress);
+
+    fn toggle_renewals(ref self: TContractState, domain: felt252, limit_price: u256);
+
+    fn renew(
+        ref self: TContractState,
+        root_domain: felt252,
+        renewer: starknet::ContractAddress,
+        limit_price: u256,
+    );
+
+    fn batch_renew(
+        ref self: TContractState,
+        domain: array::Array::<felt252>,
+        renewer: array::Array::<starknet::ContractAddress>,
+        limit_price: array::Array::<u256>,
+    );
+}
+
+#[starknet::contract]
+mod AutoRenewal {
+    use starknet::ContractAddress;
+    use starknet::{get_caller_address, get_contract_address, get_block_timestamp};
+    use traits::{TryInto, Into};
+    use option::OptionTrait;
+    use array::ArrayTrait;
+    use integer::u256_from_felt252;
+    use debug::PrintTrait;
+
+    use auto_renew_contract::interfaces::erc20::{IERC20Dispatcher, IERC20DispatcherTrait};
+    use auto_renew_contract::interfaces::naming::{INamingDispatcher, INamingDispatcherTrait};
+    use auto_renew_contract::interfaces::pricing::{IPricingDispatcher, IPricingDispatcherTrait};
+
+    #[storage]
+    struct Storage {
+        naming_contract: ContractAddress,
+        pricing_contract: ContractAddress,
+        // (renewer, domain, limit_price) -> 1 or 0
+        _is_renewing: LegacyMap::<(ContractAddress, felt252, u256), felt252>,
+        // (renewer, domain) -> timestamp
+        last_renewal: LegacyMap::<(ContractAddress, felt252), felt252>,
+    }
+
+    //
+    // Events
+    //
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        toggled_renewal: toggled_renewal,
+        domain_renewed: domain_renewed,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct toggled_renewal {
+        domain: felt252,
+        renewer: ContractAddress,
+        limit_price: u256,
+        is_renewing: felt252,
+        last_renewal: felt252,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct domain_renewed {
+        domain: felt252,
+        renewer: ContractAddress,
+        days: felt252,
+        limit_price: u256,
+        timestamp: felt252,
+    }
+
+    //
+    // Constructor
+    //
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        naming_address: ContractAddress,
+        pricing_address: ContractAddress,
+        erc20_address: ContractAddress
+    ) {
+        self.naming_contract.write(naming_address);
+        self.pricing_contract.write(pricing_address);
+
+        // approve naming contract to transfer tokens
+        let max: u256 = integer::BoundedInt::max();
+        IERC20Dispatcher { contract_address: erc20_address }.approve(naming_address, max);
```

Comment:
This was a good idea to preapprove the naming contract so spend all the funds from this contract to save on fees, but we might not pass the audit because of this. We want to guarantee our users we won't be able to spend more than their approved limit, but to do so we compare their limit to an estimation from the pricing contract. The naming contract could be upgraded to use a malicious pricing contract and "steal" more money to the users than their approved limit (that would still be checked on the old pricing). I will describe a solution which is not too expensive and allows us to avoid being dependent on the pricing contract line 223. This line can be removed.